README.md
##Keyboard
|属性|说明|类型|默认值|
|---|----|---|:----:|
|prefixCls|外部传入类|string|cefc-keyboard|
|type|键盘类型，有number、stock、letter三个值|string|number|
|showKeyboard|是否显示键盘(必需）|bool|false|
|onPress|点击操作的回调函数（必需）|func|无|
|onDelete|点击删除的回调函数|func|无|
|onHide|隐藏密码键盘后的回调函数|func|无|
|scrollHide|滑动页面则收起键盘|bool|无|
|onScroll|页面滑动回调函数|bool|无|

##使用方法
```js
<Keyboard
   type="stock"
   showKeyboard={this.state.showKeyboard}
   onPress={this.handleKeyPress}
   onDelete={this.handleKeyValueDelete}
   onHide={this.handleHideKeyboard}
   onClear={this.clearValue}
/>
```

CHANGELOG.md
## CHANGE LOG
* Keyboard组件、README初始化
* 摒弃CSSModules，样式类前面加prefixCls


src/index.js
import React, { Component, PropTypes } from 'react';
import CSSModule from 'react-css-modules';
import classnames from 'classnames';

import styles from './style/index.less';
import Tappable from './react-tappable/Tappable';

const keboardNumberItems = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0'];  //switch代表切换字母键盘的按钮
const keyboardLettersItems = {
  'firstLine': ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
  'secondLine': ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'],
  'thirdLine': ['Z', 'X', 'C', 'V', 'B', 'N', 'M']
};
const keyboardLowerCaseLettersItems = {
  firstLine: ['q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'],
  secondLine: ['a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l'],
  thirdLine: ['z', 'x', 'c', 'v', 'b', 'n', 'm']
};
const keyboardStockItems = ['600', '300', '000', '002'];
const TOUCH_MOVE = 'touchmove';

class Keyboard extends Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      isLowerCase: false,
      keyboardItems: [],
      type: this.props.type,
      showKeyboard: this.props.showKeyboard,
    };

    this.initialKeyboardType = this.props.type;
    this.keyboardTypes = { DIGIT: 'number', STOCK: 'stock', LETTER: 'letter' };
  }

  componentDidMount() {
    if (this.props.scrollHide) {
      window.addEventListener(TOUCH_MOVE, this.handleHideKeyboardTap)
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.showKeyboard !== this.state.showKeyboard) {
      this.state.showKeyboard = nextProps.showKeyboard;
      this.state.type = this.initialKeyboardType;
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    const conditions = [
      () => { return this.props.showKeyboard !== nextProps.showKeyboard },
      () => { return this.state.type !== nextState.type },
      () => { return this.state.isLowerCase !== nextState.isLowerCase }
    ];

    return conditions.some((fn) => { return fn(); });
  }

  componentWillUnmount() {
    if(this.props.scrollHide) {
      window.removeEventListener(TOUCH_MOVE, this.handleHideKeyboardTap);
    }
  }

  /**
   * 切换字母键盘大小写
   */
  handleLetterSwitchTap = () => {
    this.setState({ isLowerCase: !this.state.isLowerCase });
  }

  /**
   * 切换数字键盘
   */
  handleKeyboardSwitchTap = (keyboardType) => {
    return () => {
      this.setState({ type: keyboardType });
    }
  }

  handleHideKeyboardTap = () => {
    if (this.props.onHide) {
      this.props.onHide();
    } else {
      this.setState({ showKeyboard: !this.state.showKeyboard });
    }
  }

  handleKeyboardTouch = (value) => {
    return (e) => {
      e.preventDefault();
      e.stopPropagation(); //阻止冒泡事件，点击输入框的时候不执行body的绑定操作
      this.props.onPress(value);
    }
  }

  handleDeleteTap = () => {
    this.props.onDelete();
  }

  renderStockItems = () => {
    return (
        <ul styleName="stock-key" key="stock-key">
          {keyboardStockItems.map((item, index) => {
            return (
                <Tappable
                    component="li"
                    key={`${item}-${index}`}
                    styleName="digit-item stock-digit-item"
                    onTouchEnd={this.handleKeyboardTouch(item)}
                >
                  <div>{item}</div>
                </Tappable>
            )
          })}
          <Tappable component="li" styleName="digit-item action-bg" onTap={this.handleKeyboardSwitchTap(this.keyboardTypes.LETTER)}><div>ABC</div></Tappable>
        </ul>
    );
  }

  renderDigitKey = () => {
    const { type } = this.state;

    return (
        <ul styleName={`digit-key digit-${type}`} key="digit-key">
          {keboardNumberItems.map((item, index) => {
            return (
                <Tappable
                    component="li"
                    key={`${item}-${index}`}
                    styleName="digit-item"
                    onTouchEnd={this.handleKeyboardTouch(item)}
                >
                  <div>{item}</div>
                </Tappable>
            );
          })}
          <Tappable component="li" styleName="digit-item" onTap={this.handleHideKeyboardTap}>
            <div styleName="packUpIcon">{}</div>
          </Tappable>
        </ul>
    )
  }

  renderOperatorKey = () => {
    return (
        <ul styleName="operator-key" key="operator-key">
          <Tappable component="li" styleName="action-bg action-image" onTap={this.handleDeleteTap}>
            <div>
              <i styleName="deleteIcon" />
            </div>
          </Tappable>
          <Tappable component="li" styleName="action-bg" onTap={this.handleHideKeyboardTap}>
            <div><span>确定</span></div>
          </Tappable>
        </ul>
    );
  }

  renderLetterKeyboard = () => {
    const { isLowerCase } = this.state;
    const letterItems = isLowerCase ? keyboardLowerCaseLettersItems : keyboardLettersItems;

    return (
        <ul styleName="letterKey" key="letterKey">
          <ul styleName="flex-center">
            {letterItems.firstLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
          </ul>
          <ul styleName="flex-center letter-gap">
            {letterItems.secondLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
          </ul>
          <ul styleName="flex-center letter-gap letter-thirdLine">
            <Tappable component="div" styleName="operator-item operator-item-width action-image" onTap={this.handleLetterSwitchTap}>
              <i styleName={isLowerCase ? 'lowerCaseIcon' : 'upperCaseIcon'}>{}</i>
            </Tappable>
            {letterItems.thirdLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
            <Tappable component="div" styleName="operator-item operator-item-width action-image" onTap={this.handleDeleteTap}>
              <div styleName="deleteIcon">{}</div>
            </Tappable>
          </ul>
          <ul styleName="flex-center letter-gap letter-forthLine">
            <Tappable component="li" styleName="operator-item operator-item-width" onTap={this.handleKeyboardSwitchTap(this.keyboardTypes.DIGIT)}>123</Tappable>
            <Tappable component="li" styleName="operator-item space" onTapEnd={this.handleKeyboardTouch(' ')}>space</Tappable>
            <Tappable component="li" styleName="operator-item operator-item-width action-image" onTap={this.handleHideKeyboardTap}>
              <div styleName="packUpIcon">{}</div>
            </Tappable>
            <li styleName="operator-item confirm">确定</li>
          </ul>
        </ul>
    )
  }

  renderKeyboardItems = () => {
    const keyboardItems = [];

    switch (this.state.type) {
      case this.keyboardTypes.DIGIT:
        keyboardItems.push(this.renderDigitKey(), this.renderOperatorKey());

        break;

      case this.keyboardTypes.LETTER:
        keyboardItems.push(this.renderLetterKeyboard());

        break;

      default: //默认显示股票键盘
        keyboardItems.push(this.renderStockItems(), this.renderDigitKey(), this.renderOperatorKey());
    }

    return keyboardItems;
  }

  render() {
    const clsName = classnames({ 'hideKeyboard': !this.state.showKeyboard });

    return (
        <section styleName={`keyboard ${clsName}`} id="keyboard">
          {this.renderKeyboardItems()}
        </section>
    )
  }
}

Keyboard.propTypes = {
  type: PropTypes.oneOf(['number', 'stock', 'letter']), //标志键盘的类型
  showKeyboard: PropTypes.bool,
  onPress: PropTypes.func.isRequired,  //点击操作
  onDelete: PropTypes.func.isRequired, //删除操作
  onHide: PropTypes.func, //隐藏密码键盘后的回调函数
  scrollHide: PropTypes.bool,  //滑动页面，则收起键盘
  onScroll: PropTypes.func,    //页面滑动回调函数
};

Keyboard.defaultProps = {
  type: 'number',
  showKeyboard: false
};

export default CSSModule(Keyboard, styles, {
  allowMultiple: true
});


demo/component.js
import React from 'react';
import Keyboard from '../src/Keyboard';

class App extends React.Component {
  state = {}

  handleKeyPress = (value) => {
    console.log('handleKeyPress:', value);
  }

  handleKeyValueDelete = () => {
    console.log('handleKeyValueDelete');
  }

  handleHideKeyboard = () => {
    this.setState({
      showKeyboard: false
    });
  }

  clearValue = () => {
    console.log('clearValue');
  }

  handleToggleKeyboard = () => {
    this.setState({
      showKeyboard: !this.state.showKeyboard
    })
  }

  render() {
    return (
       <div>
         <div style={{fontSize: '.14rem'}}>Keyboard Demo</div>
         <button onClick={this.handleToggleKeyboard}>Show or hide Keyboard</button>

         <Keyboard
             type="stock"
             showKeyboard={this.state.showKeyboard}
             onPress={this.handleKeyPress}
             onDelete={this.handleKeyValueDelete}
             onHide={this.handleHideKeyboard}
             onClear={this.clearValue}
         />
       </div>

    );
  }
}

export default App;


src/react-tappable/getComponent.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var touchStyles = require('./touchStyles');
var PropTypes = require('prop-types');
var createClass = require('create-react-class');

/**
 * Tappable Component
 * ==================
 */
module.exports = function (mixins) {
	return createClass({
		displayName: 'Tappable',

		mixins: mixins,

		propTypes: {
			component: PropTypes.any, // component to create
			className: PropTypes.string, // optional className
			classBase: PropTypes.string, // base for generated classNames
			classes: PropTypes.object, // object containing the active and inactive class names
			style: PropTypes.object, // additional style properties for the component
			disabled: PropTypes.bool // only applies to buttons
		},

		getDefaultProps: function getDefaultProps() {
			return {
				component: 'span',
				classBase: 'Tappable'
			};
		},

		render: function render() {
			var props = this.props;
			var className = props.classBase + (this.state.isActive ? '-active' : '-inactive');

			if (props.className) {
				className += ' ' + props.className;
			}

			if (props.classes) {
				className += ' ' + (this.state.isActive ? props.classes.active : props.classes.inactive);
			}

			var style = {};
			_extends(style, touchStyles, props.style);

			var newComponentProps = _extends({}, props, {
				style: style,
				className: className,
				disabled: props.disabled,
				handlers: this.handlers
			}, this.handlers());

			delete newComponentProps.activeDelay;
			delete newComponentProps.classBase;
			delete newComponentProps.classes;
			delete newComponentProps.handlers;
			delete newComponentProps.onTap;
			delete newComponentProps.onPress;
			delete newComponentProps.onPinchStart;
			delete newComponentProps.onPinchMove;
			delete newComponentProps.onPinchEnd;
			delete newComponentProps.moveThreshold;
			delete newComponentProps.pressDelay;
			delete newComponentProps.pressMoveThreshold;
			delete newComponentProps.preventDefault;
			delete newComponentProps.stopPropagation;
			delete newComponentProps.component;

			return React.createElement(props.component, newComponentProps, props.children);
		}
	});
};


src/react-tappable/TappableMixin.js
'use strict';
// import PropTypes from 'prop-types';
var PropTypes = require('prop-types');
var React = require('react');
var ReactDOM = require('react-dom');

var SPACE_KEY = 32;
var ENTER_KEY = 13;

function getTouchProps(touch) {
	if (!touch) return {};
	return {
		pageX: touch.pageX,
		pageY: touch.pageY,
		clientX: touch.clientX,
		clientY: touch.clientY
	};
}

var Mixin = {
	propTypes: {
		moveThreshold: PropTypes.number, // pixels to move before cancelling tap
		activeDelay: PropTypes.number, // ms to wait before adding the `-active` class
		pressDelay: PropTypes.number, // ms to wait before detecting a press
		pressMoveThreshold: PropTypes.number, // pixels to move before cancelling press
		preventDefault: PropTypes.bool, // whether to preventDefault on all events
		stopPropagation: PropTypes.bool, // whether to stopPropagation on all events

		onTap: PropTypes.func, // fires when a tap is detected
		onPress: PropTypes.func, // fires when a press is detected
		onTouchStart: PropTypes.func, // pass-through touch event
		onTouchMove: PropTypes.func, // pass-through touch event
		onTouchEnd: PropTypes.func, // pass-through touch event
		onMouseDown: PropTypes.func, // pass-through mouse event
		onMouseUp: PropTypes.func, // pass-through mouse event
		onMouseMove: PropTypes.func, // pass-through mouse event
		onMouseOut: PropTypes.func, // pass-through mouse event
		onKeyDown: PropTypes.func, // pass-through key event
		onKeyUp: PropTypes.func },

	// pass-through key event
	getDefaultProps: function getDefaultProps() {
		return {
			activeDelay: 0,
			moveThreshold: 100,
			pressDelay: 1000,
			pressMoveThreshold: 5
		};
	},

	getInitialState: function getInitialState() {
		return {
			isActive: false,
			touchActive: false,
			pinchActive: false
		};
	},

	componentWillUnmount: function componentWillUnmount() {
		this.cleanupScrollDetection();
		this.cancelPressDetection();
		this.clearActiveTimeout();
	},

	processEvent: function processEvent(event) {
		if (this.props.preventDefault) event.preventDefault();
		if (this.props.stopPropagation) event.stopPropagation();
	},

	onTouchStart: function onTouchStart(event) {
		if (this.props.onTouchStart && this.props.onTouchStart(event) === false) return;
		this.processEvent(event);
		window._blockMouseEvents = true;
		if (event.touches.length === 1) {
			this._initialTouch = this._lastTouch = getTouchProps(event.touches[0]);
			this.initScrollDetection();
			this.initPressDetection(event, this.endTouch);
			this.initTouchmoveDetection();
			this._activeTimeout = setTimeout(this.makeActive, this.props.activeDelay);
		} else if (this.onPinchStart && (this.props.onPinchStart || this.props.onPinchMove || this.props.onPinchEnd) && event.touches.length === 2) {
			this.onPinchStart(event);
		}
	},

	makeActive: function makeActive() {
		if (!this.isMounted()) return;
		this.clearActiveTimeout();
		this.setState({
			isActive: true
		});
	},

	clearActiveTimeout: function clearActiveTimeout() {
		clearTimeout(this._activeTimeout);
		this._activeTimeout = false;
	},

	initScrollDetection: function initScrollDetection() {
		this._scrollPos = { top: 0, left: 0 };
		this._scrollParents = [];
		this._scrollParentPos = [];
		var node = ReactDOM.findDOMNode(this);

		while (node) {
			if (node.scrollHeight > node.offsetHeight || node.scrollWidth > node.offsetWidth) {
				this._scrollParents.push(node);
				this._scrollParentPos.push(node.scrollTop + node.scrollLeft);
				this._scrollPos.top += node.scrollTop;
				this._scrollPos.left += node.scrollLeft;
			}

			node = node.parentNode;
		}
	},

	initTouchmoveDetection: function initTouchmoveDetection() {
		this._touchmoveTriggeredTimes = 0;
	},

	cancelTouchmoveDetection: function cancelTouchmoveDetection() {
		if (this._touchmoveDetectionTimeout) {
			clearTimeout(this._touchmoveDetectionTimeout);
			this._touchmoveDetectionTimeout = null;
			this._touchmoveTriggeredTimes = 0;
		}
	},

	calculateMovement: function calculateMovement(touch) {
		return {
			x: Math.abs(touch.clientX - this._initialTouch.clientX),
			y: Math.abs(touch.clientY - this._initialTouch.clientY)
		};
	},

	detectScroll: function detectScroll() {
		var currentScrollPos = { top: 0, left: 0 };
		for (var i = 0; i < this._scrollParents.length; i++) {
			currentScrollPos.top += this._scrollParents[i].scrollTop;
			currentScrollPos.left += this._scrollParents[i].scrollLeft;
		}
		return !(currentScrollPos.top === this._scrollPos.top && currentScrollPos.left === this._scrollPos.left);
	},

	cleanupScrollDetection: function cleanupScrollDetection() {
		this._scrollParents = undefined;
		this._scrollPos = undefined;
	},

	initPressDetection: function initPressDetection(event, callback) {
		if (!this.props.onPress) return;
		this._pressTimeout = setTimeout((function () {
			this.props.onPress(event);
			callback();
		}).bind(this), this.props.pressDelay);
	},

	cancelPressDetection: function cancelPressDetection() {
		clearTimeout(this._pressTimeout);
	},

	onTouchMove: function onTouchMove(event) {
		if (this._initialTouch) {
			this.processEvent(event);

			if (this.detectScroll()) {
				return this.endTouch(event);
			} else {
				if (this._touchmoveTriggeredTimes++ === 0) {
					this._touchmoveDetectionTimeout = setTimeout((function () {
						if (this._touchmoveTriggeredTimes === 1) {
							this.endTouch(event);
						}
					}).bind(this), 64);
				}
			}

			this.props.onTouchMove && this.props.onTouchMove(event);
			this._lastTouch = getTouchProps(event.touches[0]);
			var movement = this.calculateMovement(this._lastTouch);
			if (movement.x > this.props.pressMoveThreshold || movement.y > this.props.pressMoveThreshold) {
				this.cancelPressDetection();
			}
			if (movement.x > this.props.moveThreshold || movement.y > this.props.moveThreshold) {
				if (this.state.isActive) {
					this.setState({
						isActive: false
					});
				} else if (this._activeTimeout) {
					this.clearActiveTimeout();
				}
			} else {
				if (!this.state.isActive && !this._activeTimeout) {
					this.setState({
						isActive: true
					});
				}
			}
		} else if (this._initialPinch && event.touches.length === 2 && this.onPinchMove) {
			this.onPinchMove(event);
			event.preventDefault();
		}
	},

	onTouchEnd: function onTouchEnd(event) {
		var _this = this;

		if (this._initialTouch) {
			this.processEvent(event);
			var afterEndTouch;
			var movement = this.calculateMovement(this._lastTouch);
			if (movement.x <= this.props.moveThreshold && movement.y <= this.props.moveThreshold && this.props.onTap) {
				event.preventDefault();
				afterEndTouch = function () {
					var finalParentScrollPos = _this._scrollParents.map(function (node) {
						return node.scrollTop + node.scrollLeft;
					});
					var stoppedMomentumScroll = _this._scrollParentPos.some(function (end, i) {
						return end !== finalParentScrollPos[i];
					});
					if (!stoppedMomentumScroll) {
						_this.props.onTap(event);
					}
				};
			}
			this.endTouch(event, afterEndTouch);
		} else if (this.onPinchEnd && this._initialPinch && event.touches.length + event.changedTouches.length === 2) {
			this.onPinchEnd(event);
			event.preventDefault();
		}
	},

	endTouch: function endTouch(event, callback) {
		this.cancelTouchmoveDetection();
		this.cancelPressDetection();
		this.clearActiveTimeout();
		if (event && this.props.onTouchEnd) {
			this.props.onTouchEnd(event);
		}
		this._initialTouch = null;
		this._lastTouch = null;
		if (callback) {
			callback();
		}
		if (this.state.isActive) {
			this.setState({
				isActive: false
			});
		}
	},

	onMouseDown: function onMouseDown(event) {
		if (window._blockMouseEvents) {
			window._blockMouseEvents = false;
			return;
		}
		if (this.props.onMouseDown && this.props.onMouseDown(event) === false) return;
		this.processEvent(event);
		this.initPressDetection(event, this.endMouseEvent);
		this._mouseDown = true;
		this.setState({
			isActive: true
		});
	},

	onMouseMove: function onMouseMove(event) {
		if (window._blockMouseEvents || !this._mouseDown) return;
		this.processEvent(event);
		this.props.onMouseMove && this.props.onMouseMove(event);
	},

	onMouseUp: function onMouseUp(event) {
		if (window._blockMouseEvents || !this._mouseDown) return;
		this.processEvent(event);
		this.props.onMouseUp && this.props.onMouseUp(event);
		this.props.onTap && this.props.onTap(event);
		this.endMouseEvent();
	},

	onMouseOut: function onMouseOut(event) {
		if (window._blockMouseEvents || !this._mouseDown) return;
		this.processEvent(event);
		this.props.onMouseOut && this.props.onMouseOut(event);
		this.endMouseEvent();
	},

	endMouseEvent: function endMouseEvent() {
		this.cancelPressDetection();
		this._mouseDown = false;
		this.setState({
			isActive: false
		});
	},

	onKeyUp: function onKeyUp(event) {
		if (!this._keyDown) return;
		this.processEvent(event);
		this.props.onKeyUp && this.props.onKeyUp(event);
		this.props.onTap && this.props.onTap(event);
		this._keyDown = false;
		this.cancelPressDetection();
		this.setState({
			isActive: false
		});
	},

	onKeyDown: function onKeyDown(event) {
		if (this.props.onKeyDown && this.props.onKeyDown(event) === false) return;
		if (event.which !== SPACE_KEY && event.which !== ENTER_KEY) return;
		if (this._keyDown) return;
		this.initPressDetection(event, this.endKeyEvent);
		this.processEvent(event);
		this._keyDown = true;
		this.setState({
			isActive: true
		});
	},

	endKeyEvent: function endKeyEvent() {
		this.cancelPressDetection();
		this._keyDown = false;
		this.setState({
			isActive: false
		});
	},

	cancelTap: function cancelTap() {
		this.endTouch();
		this._mouseDown = false;
	},

	handlers: function handlers() {
		return {
			onTouchStart: this.onTouchStart,
			onTouchMove: this.onTouchMove,
			onTouchEnd: this.onTouchEnd,
			onMouseDown: this.onMouseDown,
			onMouseUp: this.onMouseUp,
			onMouseMove: this.onMouseMove,
			onMouseOut: this.onMouseOut,
			onKeyDown: this.onKeyDown,
			onKeyUp: this.onKeyUp
		};
	}
};

module.exports = Mixin;


src/Keyboard.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './style/index.less';
import Tappable from './react-tappable/Tappable';

const keboardNumberItems = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0'];  //switch代表切换字母键盘的按钮
const keyboardLettersItems = {
  'firstLine': ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
  'secondLine': ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'],
  'thirdLine': ['Z', 'X', 'C', 'V', 'B', 'N', 'M']
};
const keyboardLowerCaseLettersItems = {
  firstLine: ['q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'],
  secondLine: ['a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l'],
  thirdLine: ['z', 'x', 'c', 'v', 'b', 'n', 'm']
};
const keyboardStockItems = ['600', '300', '000', '002'];
const TOUCH_MOVE = 'touchmove';
const defaultPrefixCls = 'cefc-keyboard';

class Keyboard extends Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      isLowerCase: false,
      keyboardItems: [],
      type: this.props.type,
      showKeyboard: this.props.showKeyboard,
    };

    this.initialKeyboardType = this.props.type;
    this.keyboardTypes = { DIGIT: 'number', STOCK: 'stock', LETTER: 'letter' };
  }

  componentDidMount() {
    if (this.props.scrollHide) {
      window.addEventListener(TOUCH_MOVE, this.handleHideKeyboardTap)
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.showKeyboard !== this.state.showKeyboard) {
      this.state.showKeyboard = nextProps.showKeyboard;
      this.state.type = this.initialKeyboardType;
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    const conditions = [
      () => { return this.props.showKeyboard !== nextProps.showKeyboard },
      () => { return this.state.type !== nextState.type },
      () => { return this.state.isLowerCase !== nextState.isLowerCase }
    ];

    return conditions.some((fn) => { return fn(); });
  }

  componentWillUnmount() {
    if(this.props.scrollHide) {
      window.removeEventListener(TOUCH_MOVE, this.handleHideKeyboardTap);
    }
  }

  /**
   * 切换字母键盘大小写
   */
  handleLetterSwitchTap = () => {
    this.setState({ isLowerCase: !this.state.isLowerCase });
  }

  /**
   * 切换数字键盘
   */
  handleKeyboardSwitchTap = (keyboardType) => {
    return () => {
      this.setState({ type: keyboardType });
    }
  }

  handleHideKeyboardTap = () => {
    if (this.props.onHide) {
      this.props.onHide();
    } else {
      this.setState({ showKeyboard: !this.state.showKeyboard });
    }
  }

  handleKeyboardTouch = (value) => {
    return (e) => {
      e.preventDefault();
      e.stopPropagation(); //阻止冒泡事件，点击输入框的时候不执行body的绑定操作
      this.props.onPress(value);
    }
  }

  handleDeleteTap = () => {
    this.props.onDelete();
  }

  renderStockItems = () => {
    const prefixCls = this.props.prefixCls;
    const stockItemsCls = classnames(`${prefixCls}-digit-item`, `${prefixCls}-stock-digit-item`);
    const actionCls = classnames(`${prefixCls}-digit-item`, `${prefixCls}-action-bg`);
    return (
        <ul className={`${prefixCls}-stock-key`} key="stock-key">
          {keyboardStockItems.map((item, index) => {
            return (
                <Tappable
                    component="li"
                    key={`${item}-${index}`}
                    className={stockItemsCls}
                    onTouchEnd={this.handleKeyboardTouch(item)}
                >
                  <div>{item}</div>
                </Tappable>
            )
          })}
          <Tappable component="li" className={actionCls} onTap={this.handleKeyboardSwitchTap(this.keyboardTypes.LETTER)}><div>ABC</div></Tappable>
        </ul>
    );
  }

  renderDigitKey = () => {
    const { type } = this.state;
    const prefixCls = this.props.prefixCls;

    return (
        <ul className={`${prefixCls}-digit-key ${prefixCls}-digit-${type}`} key="digit-key">
          {keboardNumberItems.map((item, index) => {
            return (
                <Tappable
                    component="li"
                    key={`${item}-${index}`}
                    className={`${prefixCls}-digit-item`}
                    onTouchEnd={this.handleKeyboardTouch(item)}
                >
                  <div>{item}</div>
                </Tappable>
            );
          })}
          <Tappable component="li" className={`${prefixCls}-digit-item`} onTap={this.handleHideKeyboardTap}>
            <div className={`${prefixCls}-packUpIcon`}>{}</div>
          </Tappable>
        </ul>
    )
  }

  renderOperatorKey = () => {
    const prefixCls = this.props.prefixCls;

    return (
        <ul className={`${prefixCls}-operator-key`} key="operator-key">
          <Tappable component="li" className={`${prefixCls}-action-bg ${prefixCls}-action-image`} onTap={this.handleDeleteTap}>
            <div>
              <i className={`${prefixCls}-deleteIcon`} />
            </div>
          </Tappable>
          <Tappable component="li" className={`${prefixCls}-action-bg`} onTap={this.handleHideKeyboardTap}>
            <div><span>确定</span></div>
          </Tappable>
        </ul>
    );
  }

  renderLetterKeyboard = () => {
    const { isLowerCase } = this.state;
    const prefixCls = this.props.prefixCls;
    const letterItems = isLowerCase ? keyboardLowerCaseLettersItems : keyboardLettersItems;

    return (
        <ul className={`${prefixCls}-letterKey`} key="letterKey">
          <ul className={`${prefixCls}-flex-center`}>
            {letterItems.firstLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
          </ul>
          <ul className={`${prefixCls}-flex-center ${prefixCls}-letter-gap`}>
            {letterItems.secondLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
          </ul>
          <ul className={`${prefixCls}-flex-center ${prefixCls}-letter-gap ${prefixCls}-letter-thirdLine`}>
            <Tappable component="div" className={`${prefixCls}-operator-item ${prefixCls}-operator-item-width ${prefixCls}-action-image`} onTap={this.handleLetterSwitchTap}>
              <i className={isLowerCase ? `${prefixCls}-lowerCaseIcon` : `${prefixCls}-upperCaseIcon`}>{}</i>
            </Tappable>
            {letterItems.thirdLine.map((item, index) => {
              return <Tappable component="li" key={`${item}-${index}`} onTouchEnd={this.handleKeyboardTouch(item)}><div>{item}</div></Tappable>
            })}
            <Tappable component="div" className={`${prefixCls}-operator-item ${prefixCls}-operator-item-width ${prefixCls}-action-image`} onTap={this.handleDeleteTap}>
              <div className={`${prefixCls}-deleteIcon`}>{}</div>
            </Tappable>
          </ul>
          <ul className={`${prefixCls}-flex-center ${prefixCls}-letter-gap ${prefixCls}-letter-forthLine`}>
            <Tappable component="li" className={`${prefixCls}-operator-item ${prefixCls}-operator-item-width`} onTap={this.handleKeyboardSwitchTap(this.keyboardTypes.DIGIT)}>123</Tappable>
            <Tappable component="li" className={`${prefixCls}-operator-item ${prefixCls}-space`} onTapEnd={this.handleKeyboardTouch(' ')}>space</Tappable>
            <Tappable component="li" className={`${prefixCls}-operator-item ${prefixCls}-operator-item-width ${prefixCls}-action-image`} onTap={this.handleHideKeyboardTap}>
              <div className={`${prefixCls}-packUpIcon`}>{}</div>
            </Tappable>
            <li className={`${prefixCls}-operator-item ${prefixCls}-confirm`}>确定</li>
          </ul>
        </ul>
    )
  }

  renderKeyboardItems = () => {
    const keyboardItems = [];

    switch (this.state.type) {
      case this.keyboardTypes.DIGIT:
        keyboardItems.push(this.renderDigitKey(), this.renderOperatorKey());

        break;

      case this.keyboardTypes.LETTER:
        keyboardItems.push(this.renderLetterKeyboard());

        break;

      default: //默认显示股票键盘
        keyboardItems.push(this.renderStockItems(), this.renderDigitKey(), this.renderOperatorKey());
    }

    return keyboardItems;
  }

  render() {
    const prefixCls = this.props.prefixCls;
    const clsName = classnames(prefixCls,{
        [`${prefixCls}-keyboard`]: true,
        [`${prefixCls}-hideKeyboard`]: !this.state.showKeyboard
    });

    return (
        <section className={clsName} id="keyboard">
          {this.renderKeyboardItems()}
        </section>
    )
  }
}

Keyboard.propTypes = {
  prefixCls: PropTypes.string,                          //自定义类
  type: PropTypes.oneOf(['number', 'stock', 'letter']), //标志键盘的类型
  showKeyboard: PropTypes.bool,
  onPress: PropTypes.func.isRequired,  //点击操作
  onDelete: PropTypes.func.isRequired, //删除操作
  onHide: PropTypes.func, //隐藏密码键盘后的回调函数
  scrollHide: PropTypes.bool,  //滑动页面，则收起键盘
  onScroll: PropTypes.func,    //页面滑动回调函数
};

Keyboard.defaultProps = {
  prefixCls: defaultPrefixCls,
  type: 'number',
  showKeyboard: false
};

export default Keyboard;